src/irmin/proof.ml

```diff
@@ -71,125 +89,364 @@ struct
     let t elt_t = Type.map [%typ: (H.t * elt) list] of_list to_list
   end
 
-  (* Keep track of read effects happening during a computation using
-     sets. This does not keep track of the ordering of the reads. *)
-  type read_set = { nodes : N.t Hashes.t; contents : C.t Hashes.t }
+  type mode = Produce | Serialise | Deserialise | Consume [@@deriving irmin]
+  type kind = Structural | Streamed
+
+  module Structural = struct
+    (* Keep track of read effects happening during a computation using
+       sets. This does not keep track of the ordering of the reads. *)
+    type read_set = { nodes : N.t Hashes.t; contents : C.t Hashes.t }
+    [@@deriving irmin]
+
+    type t = { mode : mode; set : read_set } [@@deriving irmin]
+
+    let empty_set () = { contents = Hashes.create 13; nodes = Hashes.create 13 }
+  end
+
+  module Streamed = struct
+    let ref_t v = Type.map v ref ( ! )
+
+    type produce = { set : unit Hashes.t; rev_elts : P.elt list ref }
+    [@@deriving irmin]
+
+    type consume = {
+      nodes : N.t Hashes.t;
+      contents : C.t Hashes.t;
+      stream : P.elt Seq.t ref;
+    }
+    [@@deriving irmin]
+
+    type t = Produce of produce | Consume of consume [@@deriving irmin]
+  end
+
+  type v = Empty | Structural of Structural.t | Streamed of Streamed.t
```

Comment:
I'm not super fond of these new names. Just keep the old one to avoid introducing new concepts:

```
type v = Empty | Set of Set.t | Stream of Stream.t
```